### PR TITLE
wibox: Add an input_passthrough property

### PR DESCRIPTION
**Pull request by Elv13**
Also improve the wibox documentation

### CODE REVIEW

**Comment by Elv13** [comment]
Umm, there seems to be an unexpected side effect in `test-struts.lua`

```
Error: running function for step 20/21 (@2): /home/travis/build/awesomeWM/awesome/tests/test-struts.lua:403: 0 + 64 == 0 called from line 513

stack traceback:

	[C]: in function 'assert'

	/home/travis/build/awesomeWM/awesome/tests/test-struts.lua:403: in function 'test_workarea'

	/home/travis/build/awesomeWM/awesome/tests/test-struts.lua:513: in function </home/travis/build/awesomeWM/awesome/tests/test-struts.lua:475>

	(...tail calls...)

	[C]: in function 'xpcall'

	/home/travis/build/awesomeWM/awesome/tests/_runner.lua:73: in function </home/travis/build/awesomeWM/awesome/tests/_runner.lua:64>

	[C]: in function 'xpcall'

	lib/gears/protected_call.lua:36: in function <lib/gears/protected_call.lua:35>

	(...tail calls...)

	lib/gears/timer.lua:231: in function <lib/gears/timer.lua:229>!

===> ERROR running /home/travis/build/awesomeWM/awesome/tests/test-struts.lua <===

Error: running function for step 20/21 (@2): /home/travis/build/awesomeWM/awesome/tests/test-struts.lua:403: 0 + 64 == 0 called from line 513
```

It doesn't seem to happen locally.

File: docs/common/wibox.ldoc [hunk 1]
```diff
@@ -180,6 +180,15 @@
 -- @property shape
 -- @tparam gears.shape shape
 
+--- Forward the inputs to the client below the wibox.
```
Comment (psychon) [inline docs/common/wibox.ldoc]: I would mention that this is implemented by setting `shape_input` so that it becomes clearer that the two properties interact.

File: lib/wibox/init.lua [hunk 2]
```diff
@@ -137,10 +137,23 @@ function wibox:_apply_shape()
...
+    if not self._input_passthrough then
+        self.shape_input = nil
+        return
+    end
+
+    local geo = self:geometry()
+    local bw = self.border_width
+
+    -- First handle the bounding shape (things including the border)
+    local img = cairo.ImageSurface(cairo.Format.A1, geo.width + 2*bw, geo.height + 2*bw)
```
Comment (psychon) [inline lib/wibox/init.lua]: Is it enough to use size `0x0` for the image surface? I would think so.

File: lib/wibox/init.lua [hunk 3]
```diff
@@ -137,10 +137,23 @@ function wibox:_apply_shape()
...
+function wibox:_apply_input_shape()
+    if not self._input_passthrough then
+        self.shape_input = nil
+        return
+    end
+
+    local geo = self:geometry()
+    local bw = self.border_width
+
+    -- First handle the bounding shape (things including the border)
```
Comment (psychon) [inline lib/wibox/init.lua]: Copy&paste

File: lib/wibox/init.lua [hunk 4]
```diff
@@ -151,6 +164,16 @@ function wibox:get_shape()
     return self._shape
 end
 
+function wibox:set_input_passthrough(value)
+    rawset(self, "_input_passthrough", value)
+    self:_apply_input_shape()
```
Comment (psychon) [inline lib/wibox/init.lua]: Why the separate function? Just do
```lua
if value then
    local img = cairo.ImageSurface(cairo.Format.A1, 0, 0)
    self.shape_inut = img._native
    img:finish()
else
    self.shape_input = nil
end
rawset(self, "_input_passthrough", value)
self:emit_signal("property::input_passthrough", value)
```

File: lib/wibox/init.lua [hunk 5]
```diff
@@ -325,6 +348,7 @@ local function new(args)
     end
 
     ret.shape = args.shape
+    ret.input_passthrough = args.input_passthrough
```
Comment (psychon) [inline lib/wibox/init.lua]: I'd wrap this in an `if args.input_passthrough then`.

File: tests/test-wibox-shape.lua [hunk 6]
```diff
@@ -18,14 +18,103 @@ local wb = wibox {
...
     end
-})
+}
+
+local function click(x, y)
+    table.insert(steps, function()
+        mouse.coords{x=x, y=y}
+        return true
+    end)
+    table.insert(steps, function()
```
Comment (psychon) [inline tests/test-wibox-shape.lua]: You can merge these two into one function, I think. At least I do not see why not.

File: tests/test-wibox-shape.lua [hunk 7]
```diff
@@ -18,14 +18,103 @@ local wb = wibox {
...
+    end)
+end
+
+-- Check each corner
+click(0 , 100)
+click(99, 100)
+click(99, 199)
+click(0 , 199)
+
+-- Batch the count check because they fail a couple time each
```
Comment (psychon) [inline tests/test-wibox-shape.lua]: Just add a call to `awesome.sync()` after all the clicks and one "idle loop" of the test runner afterwards (= `check_count` must stay in a separate step, the other steps sending the clicks can be merged) and this should not need a loop anymore.

File: tests/test-wibox-shape.lua [hunk 8]
```diff
@@ -18,14 +18,103 @@ local wb = wibox {
...
+check_count(4)
+
+table.insert(steps, function()
+    wb.input_passthrough = true
+    count = 0
+    awesome.sync()
```
Comment (psychon) [inline tests/test-wibox-shape.lua]: This sync should not be necessary. I assume it is for making sure that `input_passthrough` was handled by the X server, but that should not be necessary. I think you just need a `sync()` when you wait for events generated by the X server.

File: tests/test-wibox-shape.lua [hunk 9]
```diff
@@ -18,14 +18,103 @@ local wb = wibox {
...
+end)
+
+
+-- Do it again
+click(0 , 100)
+click(99, 100)
+click(99, 199)
+click(0 , 199)
+
+-- Batch the count check because they fail a couple time each
```
Comment (psychon) [inline tests/test-wibox-shape.lua]: Same as above.

File: tests/test-wibox-shape.lua [hunk 10]
```diff
@@ -18,14 +18,103 @@ local wb = wibox {
...
+check_count(0)
+
+table.insert(steps, function()
+    wb.input_passthrough = false
+    count = 0
+    return true
+end)
+
+-- It takes a while for the shape to be applied, waste time
```
Comment (psychon) [inline tests/test-wibox-shape.lua]: No it does not. Well, okay, it does, but that should not be the problem here. The only thing that needs a `sync()` is when you fake input and the next step of the test expects that input to be handled already.

File: lib/awful/wibar.lua [hunk 11]
```diff
@@ -339,37 +339,19 @@ end
...
+-- @tparam string args.position The position.
+-- @tparam string args.stretch If the wibar need to be stretched to fill the screen.
+--@DOC_wibox_constructor_COMMON@
 -- @return The new wibar
 -- @function awful.wibar
-function awfulwibar.new(arg)
-    arg = arg or {}
-    local position = arg.position or "top"
+function awfulwibar.new(args)
+    args = args or {}
```
Comment (psychon) [inline lib/awful/wibar.lua]: This seems a lot like an (almost) unrelated change that should be in its own commit, but okay, whatever....

File: docs/load_ldoc.cmake [hunk 12]
```diff
@@ -13,8 +13,11 @@ foreach(doc_file_name ${doc_files})
     # Remove the file extension
     string(REGEX REPLACE "\\.ldoc" "" DOC_FILE_NAME ${doc_file_name})
 
+    # There is a trailing \n, remove it or it cannot be included in existing blocks
+    string(REGEX REPLACE "\n$" "" doc_file_content "${doc_file_content}")
+
     # Create a new variable usable from lua files
-    set(DOC_${DOC_FILE_NAME}_COMMON "Imported documentation\n\n${doc_file_content}")
+    set(DOC_${DOC_FILE_NAME}_COMMON "${doc_file_content}")
```
Comment (psychon) [inline docs/load_ldoc.cmake]: This should cause lots of changes in the docs. Is this really okay? (I have honestly no idea)
At least the commit message could mention/explain this... (and also why it is needed)

Comment (Elv13) [reply docs/load_ldoc.cmake]: Its necessary to be able to do

```
--- Foobar.
-- @tparam table args
-- @tparam args.new_property
-- @DOC_wibox_constructor_COMMON@
-- @treturn boolean
function new(args) end
```

Without this, the previous result was

```
--- Foobar.
-- @tparam table args
-- @tparam args.new_property

-- Imported documentation

-- @tparam args.shared_property

-- @treturn boolean
function new(args) end
```

instead of :

```
--- Foobar.
-- @tparam table args
-- @tparam args.new_property
-- @tparam args.shared_property
-- @treturn boolean
function new(args) end
```

It doesn't change anything for existing doc beside the lack of newlines as long as there is a new line between the macros

Ok:

```
--@DOC_widget_COMMON@

--@DOC_object_COMMON@
```

Potentially not ok, but nevertheless mostly ok:

```
--@DOC_widget_COMMON@
--@DOC_object_COMMON@
```

**Comment by Elv13** [comment]
All done

File: docs/common/wibox.ldoc [hunk 13]
```diff
@@ -180,6 +180,22 @@
...
+--
+-- This replace the `shape_input` mask with an empty area. All mouse and
+-- keyboard events are sent to the object (such as a client) positionned below
```
Comment (psychon) [inline docs/common/wibox.ldoc]: `positionned` -> `positioned`

File: lib/wibox/init.lua [hunk 14]
```diff
@@ -137,11 +137,6 @@ function wibox:_apply_shape()
     img:finish()
 end
 
---- Set the wibox shape.
--- @property shape
--- @tparam gears.shape A gears.shape compatible function.
--- @see gears.shape
```
Comment (psychon) [inline lib/wibox/init.lua]: Huh? Why is this removed?

File: tests/test-wibox-shape.lua [hunk 15]
```diff
@@ -18,14 +18,107 @@ local wb = wibox {
...
+            root.fake_input("button_press", 1)
+        end
+
+        return true
+    end)
+end
+
+local function check_count(cnt)
+    table.insert(steps, function()
+        awesome.sync()
```
Comment (psychon) [inline tests/test-wibox-shape.lua]: I think this `sync` should be before the `return true` in `click` and then this function can just `return cnt == count`, but feel free to ignore me here....

File: lib/wibox/init.lua (hunk 14 above)
Comment (Elv13) [reply lib/wibox/init.lua]: It's in `docs/common/wibox.ldoc` too
